File: src/lib/ebike/range.ts
/**
 * F-EBK-004 Reichweitenprognose — Physik + P1 Selbstkalibrierung
 *
 * P = (Crr·m·g·cosα + ½ρ·CdA·v² + m·g·sinα) · v
 * Wh/km = P_motor / v  (mit η_antrieb)
 *
 * Quellen:
 * - Spec-Formel F-EBK-004
 * - ebikes.ca / Kreuzotter: Crr ~0.004–0.012, CdA ~0.3–0.8
 * - DinCalculator E-bike Route: flach ~8–14 Wh/km, 6 % ~18–25 Wh/km
 * - Garage-Kopplung: Reifendruck verändert Crr messbar
 */

import type { Bike, Ride, RiderProfile } from "@/types";

export interface RangeCalibration {
  crr: number;
  cdA: number;
  riderPowerW: number;
  samples: number;
  updatedAt: string;
}

export interface RangeEstimate {
  kmLow: number;
  kmHigh: number;
  whPerKmLow: number;
  whPerKmHigh: number;
  batteryWh: number;
  confidence: "low" | "medium" | "high";
  factors: string[];
  calibrated: boolean;
}

const G = 9.81;
const RHO = 1.225;

function batteryWh(bike: Bike): number {
  const bat = bike.components.find(
    (c) => c.slot === "battery" && !c.removedAt
  );
  const fromModel = bat?.componentModelId;
  // PowerTube 800 etc. in catalog attributes — fallback
  if (fromModel === "cm-bosch-powertube-800") return 800;
  const attr = bat?.attributes.find((a) => a.key === "capacity_wh");
  return attr?.valueNum ?? 500;
}

function tirePressurePsi(bike: Bike): number {
  const t = bike.components.find((c) => c.slot === "tire_rear" && !c.removedAt);
  const p = t?.currentSettings.pressure_psi;
  return typeof p === "number" ? p : 24;
}

function baseCrr(bike: Bike): number {
  // Stollenreifen / niedriger Druck → höherer Crr (Drewanz / Praxis)
  const psi = tirePressurePsi(bike);
  let crr = bike.category === "emtb" || bike.category === "mtb_enduro" ? 0.018 : 0.012;
  if (psi < 22) crr += 0.004;
  else if (psi > 28) crr -= 0.002;
  return crr;
}

function baseCdA(category: Bike["category"]): number {
  // Unimore defaults Cd≈0.7, A≈0.5 → CdA≈0.35; MTB aufrechter
  switch (category) {
    case "road":
      return 0.32;
    case "gravel":
      return 0.36;
    case "emtb":
    case "mtb_enduro":
    case "mtb_am":
      return 0.42;
    default:
      return 0.38;
  }
}

export function defaultCalibration(
  bike: Bike,
  profile: RiderProfile
): RangeCalibration {
  return {
    crr: baseCrr(bike),
    cdA: baseCdA(bike.category),
    riderPowerW: 90 + profile.skillLevel * 15,
    samples: 0,
    updatedAt: new Date().toISOString(),
  };
}

/** Kalman-artige Ein-Schritt-Anpassung nach Ride */
export function calibrateFromRide(
  prev: RangeCalibration,
  ride: Ride,
  batteryWhUsed: number
): RangeCalibration {
  if (ride.distanceM < 2000 || batteryWhUsed < 10) return prev;
  const km = ride.distanceM / 1000;
  const observedWhPerKm = batteryWhUsed / km;
  // Ziel: Wh/km ≈ f(crr, cdA) — grobe Inversion über Mischfaktor
  const expected =
    prev.crr * 180 + prev.cdA * 40 + 200 / Math.max(8, ride.distanceM / ride.durationSec * 3.6);
  const err = observedWhPerKm - expected;
  const k = 1 / (prev.samples + 2);
  return {
    crr: Math.min(0.04, Math.max(0.006, prev.crr + k * err * 0.00008)),
    cdA: Math.min(0.6, Math.max(0.25, prev.cdA + k * err * 0.002)),
    riderPowerW: Math.min(
      250,
      Math.max(60, prev.riderPowerW + k * ((ride.motorData?.avgRiderPower ?? prev.riderPowerW) - prev.riderPowerW))
    ),
    samples: prev.samples + 1,
    updatedAt: new Date().toISOString(),
  };
}

export function estimateRange(input: {
  bike: Bike;
  profile: RiderProfile;
  calibration?: RangeCalibration;
  /** mittlere Steigung rad */
  meanGrade?: number;
  speedKmh?: number;
  socPercent?: number;
}): RangeEstimate {
  const cal = input.calibration ?? defaultCalibration(input.bike, input.profile);
  const m =
    (input.profile.riderWeightKg ?? 78) +
    (input.bike.weightKg ?? 15) +
    2; // leichtes Gepäck
  const v = (input.speedKmh ?? 22) / 3.6;
  const alpha = input.meanGrade ?? 0.03;
  const crr = cal.crr;
  const cdA = cal.cdA;

  const pRoll = crr * m * G * Math.cos(alpha) * v;
  const pAero = 0.5 * RHO * cdA * v ** 3;
  const pGrade = m * G * Math.sin(alpha) * v;
  const pTotal = pRoll + pAero + pGrade;
  const pRider = Math.min(cal.riderPowerW, pTotal * 0.55);
  const eta = 0.78;
  const pMotor = Math.max(0, (pTotal - pRider) / eta);

  const whPerKm = (pMotor / v) * (1000 / 3600); // W·s/m → Wh/km
  const wh = batteryWh(input.bike) * ((input.socPercent ?? 100) / 100);

  // Spanne: ±18 % Unsicherheit, enger mit mehr Samples
  const spread = Math.max(0.08, 0.18 - cal.samples * 0.015);
  const whLow = whPerKm * (1 - spread);
  const whHigh = whPerKm * (1 + spread);

  const factors = [
    `Crr ${crr.toFixed(3)} (Reifendruck ${tirePressurePsi(input.bike)} psi)`,
    `Cd·A ${cdA.toFixed(2)} (${input.bike.category})`,
    `Systemmasse ${m.toFixed(0)} kg · Ø-Steigung ${(alpha * 100).toFixed(1)} %`,
  ];

  return {
    kmLow: Math.round(wh / whHigh),
    kmHigh: Math.round(wh / whLow),
    whPerKmLow: Math.round(whLow * 10) / 10,
    whPerKmHigh: Math.round(whHigh * 10) / 10,
    batteryWh: Math.round(wh),
    confidence:
      cal.samples >= 5 ? "high" : cal.samples >= 2 ? "medium" : "low",
    factors,
    calibrated: cal.samples > 0,
  };
}

File: mobile/lib/domain/ebike/range.dart
/// Vereinfachte F-EBK-004 Reichweitenprognose — Port von src/lib/ebike/range.ts.
library;

import 'dart:math' as math;

import '../bike.dart';

const _g = 9.81;
const _rho = 1.225;

class RangeCalibration {
  const RangeCalibration({
    required this.crr,
    required this.cdA,
    required this.riderPowerW,
    this.samples = 0,
    this.updatedAt,
  });

  final double crr;
  final double cdA;
  final double riderPowerW;
  final int samples;
  final String? updatedAt;

  Map<String, dynamic> toJson() => {
        'crr': crr,
        'cdA': cdA,
        'riderPowerW': riderPowerW,
        'samples': samples,
        if (updatedAt != null) 'updatedAt': updatedAt,
      };

  factory RangeCalibration.fromJson(Map<String, dynamic>? json) {
    if (json == null) {
      return const RangeCalibration(crr: 0.015, cdA: 0.4, riderPowerW: 120);
    }
    return RangeCalibration(
      crr: (json['crr'] as num?)?.toDouble() ?? 0.015,
      cdA: (json['cdA'] as num?)?.toDouble() ?? 0.4,
      riderPowerW: (json['riderPowerW'] as num?)?.toDouble() ?? 120,
      samples: (json['samples'] as num?)?.toInt() ?? 0,
      updatedAt: json['updatedAt'] as String?,
    );
  }
}

class RangeEstimate {
  const RangeEstimate({
    required this.kmLow,
    required this.kmHigh,
    required this.whPerKmLow,
    required this.whPerKmHigh,
    required this.batteryWh,
    required this.confidence,
    required this.factors,
    required this.calibrated,
  });

  final int kmLow;
  final int kmHigh;
  final double whPerKmLow;
  final double whPerKmHigh;
  final int batteryWh;
  final String confidence; // low | medium | high
  final List<String> factors;
  final bool calibrated;
}

double _baseCrr(BikeCategory category, double tirePressurePsi) {
  var crr =
      category == BikeCategory.emtb || category == BikeCategory.mtbEnduro
          ? 0.018
          : 0.012;
  if (tirePressurePsi < 22) {
    crr += 0.004;
  } else if (tirePressurePsi > 28) {
    crr -= 0.002;
  }
  return crr;
}

double _baseCdA(BikeCategory category) {
  switch (category) {
    case BikeCategory.road:
      return 0.32;
    case BikeCategory.gravel:
      return 0.36;
    case BikeCategory.emtb:
    case BikeCategory.mtbEnduro:
    case BikeCategory.mtbAm:
      return 0.42;
    default:
      return 0.38;
  }
}

RangeCalibration defaultCalibration({
  required BikeCategory category,
  double tirePressurePsi = 24,
  int skillLevel = 2,
}) {
  return RangeCalibration(
    crr: _baseCrr(category, tirePressurePsi),
    cdA: _baseCdA(category),
    riderPowerW: 90 + skillLevel * 15.0,
  );
}

/// Kalman-artige Ein-Schritt-Anpassung nach Ride (Port calibrateFromRide).
RangeCalibration calibrateFromRide({
  required RangeCalibration prev,
  required double distanceKm,
  required double movingTimeSec,
  required double batteryWhUsed,
  double? avgRiderPowerW,
}) {
  if (distanceKm < 2 || batteryWhUsed < 10) return prev;
  final observedWhPerKm = batteryWhUsed / distanceKm;
  final speedKmh =
      movingTimeSec > 0 ? distanceKm / (movingTimeSec / 3600) : 18;
  final expected =
      prev.crr * 180 + prev.cdA * 40 + 200 / math.max(8, speedKmh);
  final err = observedWhPerKm - expected;
  final k = 1 / (prev.samples + 2);
  final nextPower = avgRiderPowerW ?? prev.riderPowerW;
  return RangeCalibration(
    crr: (prev.crr + k * err * 0.00008).clamp(0.006, 0.04),
    cdA: (prev.cdA + k * err * 0.002).clamp(0.25, 0.6),
    riderPowerW:
        (prev.riderPowerW + k * (nextPower - prev.riderPowerW)).clamp(60, 250),
    samples: prev.samples + 1,
    updatedAt: DateTime.now().toUtc().toIso8601String(),
  );
}

/// Physik: P = (Crr·m·g·cosα + ½ρ·CdA·v² + m·g·sinα) · v
RangeEstimate estimateRange({
  required BikeCategory category,
  RangeCalibration? calibration,
  double batteryWh = 500,
  double tirePressurePsi = 24,
  double riderWeightKg = 78,
  double bikeWeightKg = 22,
  double meanGrade = 0.03,
  double speedKmh = 22,
  double socPercent = 100,
  int skillLevel = 2,
}) {
  final cal = calibration ??
      defaultCalibration(
        category: category,
        tirePressurePsi: tirePressurePsi,
        skillLevel: skillLevel,
      );
  final m = riderWeightKg + bikeWeightKg + 2;
  final v = speedKmh / 3.6;
  final alpha = meanGrade;
  final crr = cal.crr;
  final cdA = cal.cdA;

  final pRoll = crr * m * _g * math.cos(alpha) * v;
  final pAero = 0.5 * _rho * cdA * v * v * v;
  final pGrade = m * _g * math.sin(alpha) * v;
  final pTotal = pRoll + pAero + pGrade;
  final pRider =
      cal.riderPowerW < pTotal * 0.55 ? cal.riderPowerW : pTotal * 0.55;
  const eta = 0.78;
  final pMotor = math.max(0.0, (pTotal - pRider) / eta);

  final whPerKm = (pMotor / v) * (1000 / 3600);
  final wh = batteryWh * (socPercent / 100);

  final spread = math.max(0.08, 0.18 - cal.samples * 0.015);
  final whLow = whPerKm * (1 - spread);
  final whHigh = whPerKm * (1 + spread);

  return RangeEstimate(
    kmLow: (wh / whHigh).round(),
    kmHigh: (wh / whLow).round(),
    whPerKmLow: (whLow * 10).round() / 10,
    whPerKmHigh: (whHigh * 10).round() / 10,
    batteryWh: wh.round(),
    confidence: cal.samples >= 5
        ? 'high'
        : cal.samples >= 2
            ? 'medium'
            : 'low',
    factors: [
      'Crr ${crr.toStringAsFixed(3)} (Reifendruck ${tirePressurePsi.toStringAsFixed(0)} psi)',
      'Cd·A ${cdA.toStringAsFixed(2)} (${category.name})',
      'Systemmasse ${m.toStringAsFixed(0)} kg · Ø-Steigung ${(alpha * 100).toStringAsFixed(1)} %',
    ],
    calibrated: cal.samples > 0,
  );
}
